service/udp_test.go

```diff
@@ -140,26 +146,74 @@ func TestIPFilter(t *testing.T) {
 			},
 			payload: ciphertext,
 		}
-
-		service.GracefulStop()
-		return metrics
 	}
 
+	service.GracefulStop()
+	return metrics
+}
+
+func TestIPFilter(t *testing.T) {
+	// Test both the first-packet and subsequent-packet cases.
+	payloads := [][]byte{[]byte("payload1"), []byte("payload2")}
+
 	t.Run("Localhost allowed", func(t *testing.T) {
-		metrics := checkLocalhost(allowAll)
+		metrics := sendToDiscard(payloads, allowAll)
 		if metrics.natEntriesAdded != 1 {
 			t.Errorf("Expected 1 NAT entry, not %d", metrics.natEntriesAdded)
 		}
 	})
 
 	t.Run("Localhost not allowed", func(t *testing.T) {
-		metrics := checkLocalhost(onet.RequirePublicIP)
+		metrics := sendToDiscard(payloads, onet.RequirePublicIP)
 		if metrics.natEntriesAdded != 0 {
 			t.Error("Unexpected NAT entry on rejected packet")
 		}
+		if len(metrics.upstreamPackets) != 2 {
+			t.Errorf("Expected 2 reports, not %v", metrics.upstreamPackets)
+		}
+		for _, report := range metrics.upstreamPackets {
+			if report.clientProxyBytes == 0 {
```

Comment:
You may want to use testify to make the assertions more concise.

Reply:
Done